src/openai/_tracing.py

```diff
@@ -0,0 +1,160 @@
+from __future__ import annotations
+
+import json
+import functools
+from typing import TypeVar, Callable, Any, Generator, Iterator, Union, Literal
+from typing_extensions import ParamSpec
+
+from ._extras import opentelemetry as trace, has_tracing_enabled
+from .types.chat import ChatCompletion, ChatCompletionChunk
+from .types.completion import Completion
+from ._streaming import Stream
+
+TracedMessageTypes = Literal["system", "user", "assistant", "tool"]
+TracedModels = Union[ChatCompletion, Completion]
+
+_P = ParamSpec("_P")
+_R = TypeVar("_R")
+
+
+def _set_attribute(span: trace.Span, key: str, value: Any) -> None:
+    if value:
+        span.set_attribute(key, value)
+
+
+def _add_request_chat_message_event(span: trace.Span, **kwargs: Any) -> None:
+    for message in kwargs.get("messages", []):
+        try:
+            message = message.to_dict()
+        except AttributeError:
+            pass
+        
+        if message.get("role") in TracedMessageTypes:
+            name = f"gen_ai.{message.get('role')}.message"
+            span.add_event(
```

Comment:
Question: are events ordered in OTEL? Messages are.

Reply:
For posterity, from [the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#add-events):

> Events SHOULD preserve the order in which they are recorded. This will typically match the ordering of the events' timestamps, but events may be recorded out-of-order using custom timestamps.

The wording is a bit inconsistent in the spec sometimes referring to events as a "set" (which usually implies that they are not ordered) in which case I would have had some concerns about multiple events being recorded at the same millisecond (which is the lowest resolution for timestamps in OTEL) causing jumbling of message ordering.